File: packages/gavio-py/gavio/testing/harness.py
"""GavioTestKit — run interceptor chains in isolation for unit tests.

Full version (F-DX-03) ships in v0.2.0; this v0.1.0 kit already lets you drive
a chain against a :class:`MockProvider` and assert on PII detection, the
redacted request, and the resulting audit record.
"""

from __future__ import annotations

from ..context import InterceptorContext
from ..interceptors.base import Interceptor
from ..interceptors.chain import Executor, InterceptorChain
from ..interceptors.reliability.policy import ExecutorPolicy
from ..providers.base import ProviderAdapter
from ..providers.mock import MockProvider
from ..request import GavioRequest
from ..response import GavioResponse
from ..types import Message, Provider


class _CaptureInterceptor(Interceptor):
    """Records the request as it reaches the provider (post-redaction)."""

    def __init__(self) -> None:
        self.captured: GavioRequest | None = None

    @property
    def name(self) -> str:
        return "_capture"

    async def before(
        self, request: GavioRequest, ctx: InterceptorContext
    ) -> GavioRequest:
        self.captured = request
        return request


class GavioTestKit:
    def __init__(
        self,
        interceptors: list[Interceptor] | None = None,
        provider: ProviderAdapter | None = None,
        model: str = "mock",
    ) -> None:
        self._interceptors = list(interceptors or [])
        self._provider = provider or MockProvider()
        self._model = model
        self._capture = _CaptureInterceptor()
        self.ctx: InterceptorContext | None = None
        self.response: GavioResponse | None = None

    async def run(self, messages: list[Message], **options: object) -> GavioResponse:
        request = GavioRequest(
            messages=messages,
            model=self._model,
            provider=Provider.coerce(self._provider.provider_name),
            options=dict(options),
        )
        ctx = InterceptorContext(trace_id=request.trace_id)

        all_interceptors = [*self._interceptors, self._capture]
        policies = [i for i in all_interceptors if isinstance(i, ExecutorPolicy)]
        regular = [i for i in all_interceptors if not isinstance(i, ExecutorPolicy)]
        chain = InterceptorChain(regular)

        async def base(req: GavioRequest) -> GavioResponse:
            return await self._provider.complete(req)

        executor: Executor = base
        for policy in reversed(policies):
            executor = self._wrap(policy, executor, ctx)

        self.response = await chain.execute(request, ctx, executor)
        self.ctx = ctx
        return self.response

    @staticmethod
    def _wrap(policy: ExecutorPolicy, inner: Executor, ctx: InterceptorContext):
        async def wrapped(req: GavioRequest) -> GavioResponse:
            return await policy.around(req, ctx, inner)

        return wrapped

    # ---- assertions / inspection -------------------------------------------

    @property
    def redacted_request(self) -> GavioRequest | None:
        return self._capture.captured

    @property
    def audit_record(self):
        return self.response.audit if self.response else None

    def pii_detected(self, entity_type: str | None = None) -> bool:
        if self.ctx is None:
            return False
        if entity_type is None:
            return bool(self.ctx.pii_entity_types)
        return entity_type in self.ctx.pii_entity_types

File: packages/gavio-js/src/testing/harness.ts
/**
 * GavioTestKit — run interceptor chains in isolation for unit tests.
 *
 * This v0.1.0 kit drives a chain against a {@link mockProvider} and lets you
 * assert on PII detection, the redacted request, and the resulting audit record.
 */

import { InterceptorContext } from '../context.js'
import { isExecutorPolicy } from '../interceptors/base.js'
import type { Executor, ExecutorPolicy, Interceptor } from '../interceptors/base.js'
import { InterceptorChain } from '../interceptors/chain.js'
import type { AuditRecord } from '../interceptors/audit/record.js'
import { mockProvider } from '../providers/mock.js'
import type { ProviderAdapter } from '../providers/base.js'
import { GavioRequest } from '../request.js'
import type { GavioResponse } from '../response.js'
import { coerceProvider } from '../types.js'
import type { Message } from '../types.js'

/** Records the request as it reaches the provider (post-redaction). */
class CaptureInterceptor implements Interceptor {
  readonly name = '_capture'
  captured: GavioRequest | null = null

  before(request: GavioRequest): GavioRequest {
    this.captured = request
    return request
  }
}

export interface GavioTestKitOptions {
  interceptors?: Interceptor[]
  provider?: ProviderAdapter
  model?: string
}

export interface RunResult {
  response: GavioResponse
  ctx: InterceptorContext
  /** The request text as it reached the provider (post-redaction). */
  preRequestText(): string
  /** True if any (or a specific) PII entity type was detected. */
  piiDetected(entityType?: string): boolean
  /** The audit record produced for this call, if an audit interceptor ran. */
  readonly auditRecord: AuditRecord | null
}

export class GavioTestKit {
  private readonly interceptors: Interceptor[]
  private readonly provider: ProviderAdapter
  private readonly model: string

  constructor(options: GavioTestKitOptions = {}) {
    this.interceptors = [...(options.interceptors ?? [])]
    this.provider = options.provider ?? mockProvider()
    this.model = options.model ?? 'mock'
  }

  async run(input: { messages: Message[]; options?: Record<string, unknown> }): Promise<RunResult> {
    const request = new GavioRequest({
      messages: input.messages,
      model: this.model,
      provider: coerceProvider(this.provider.providerName),
      options: input.options ?? {},
    })
    const ctx = new InterceptorContext({ traceId: request.traceId })

    const capture = new CaptureInterceptor()
    const all = [...this.interceptors, capture]
    const policies = all.filter(isExecutorPolicy)
    const regular = all.filter((i) => !isExecutorPolicy(i))
    const chain = new InterceptorChain(regular)

    let executor: Executor = (req) => this.provider.complete(req)
    for (let i = policies.length - 1; i >= 0; i--) {
      executor = wrap(policies[i]!, executor, ctx)
    }

    const response = await chain.execute(request, ctx, executor)

    return {
      response,
      ctx,
      preRequestText: () => capture.captured?.promptText() ?? '',
      piiDetected: (entityType?: string): boolean => {
        if (entityType === undefined) return ctx.piiEntityTypes.length > 0
        return ctx.piiEntityTypes.includes(entityType)
      },
      get auditRecord(): AuditRecord | null {
        return response.audit
      },
    }
  }
}

function wrap(policy: ExecutorPolicy, inner: Executor, ctx: InterceptorContext): Executor {
  return (req: GavioRequest) => policy.around(req, ctx, inner)
}
